### PR TITLE
Add an API to notify the BuildEngine/BuildSystem delegate when it is determined that a rule needs to run

### PR DESCRIPTION
**Pull request by owenv**
…, and why

For example, because it has never been built, or because one of its inputs was rebuilt

rdar://95208406

### CODE REVIEW

**Comment by owenv** [comment]
@swift-ci please test

File: products/libllbuild/include/llbuild/buildsystem.h [hunk 1]
```diff
@@ -203,6 +203,24 @@ typedef enum LLBUILD_ENUM_ATTRIBUTES {
...
+    llb_rule_run_reason_signature_changed LLBUILD_SWIFT_NAME(signatureChanged) = 1,
+
+    /// The rule has an invalid value.
+    llb_rule_run_reason_invalid_value LLBUILD_SWIFT_NAME(invalidValue) = 2,
+
+    /// An input of the rule has been rebuilt.
+    llb_rule_run_reason_input_rebuilt LLBUILD_SWIFT_NAME(inputRebuilt) = 3,
+
+    /// The rule has never been built.
+    llb_rule_run_reason_forced = 4,
```
Comment (dmbryson) [inline products/libllbuild/include/llbuild/buildsystem.h:221]: LLBUILD_SWIFT_NAME for consistency?

**Review by dmbryson — APPROVED** [review]

File: products/libllbuild/BuildSystem-C-API.cpp [hunk 2]
```diff
@@ -471,6 +471,39 @@ class CAPIBuildSystemFrontendDelegate : public BuildSystemFrontendDelegate {
     uint64_t count() { return rules.size(); }
   };
 
+  static llb_rule_run_reason_t
+  convertRunReason(core::Rule::RunReason reason) {
```
Comment (BenchR267) [inline products/libllbuild/BuildSystem-C-API.cpp:475]: Maybe ad a commend to both types and do a static cast? They're based on the same raw value.

File: lib/Core/BuildEngine.cpp [hunk 3]
```diff
@@ -479,6 +483,7 @@ class BuildEngineImpl : public BuildDBDelegate {
       if (trace)
         trace->ruleNeedsToRunBecauseInvalidValue(ruleInfo.rule.get());
```
Comment (BenchR267) [inline lib/Core/BuildEngine.cpp:484]: Would be cool if the trace would also use the new enum

**Review by BenchR267 — APPROVED** [review]

File: products/libllbuild/BuildSystem-C-API.cpp (hunk 2 above)
Comment (owenv) [reply products/libllbuild/BuildSystem-C-API.cpp:475]: I think I prefer the current approach to maintain consistency with the other enums, so we have a clearer boundary between the C & C++ APIs

File: lib/Core/BuildEngine.cpp (hunk 3 above)
Comment (owenv) [reply lib/Core/BuildEngine.cpp:484]: That's a good idea. I'll file a follow-up bug for the refactor

**Comment by owenv** [comment]
@swift-ci please smoke test

File: products/libllbuild/BuildSystem-C-API.cpp (hunk 2 above)
Comment (jakepetroules) [reply products/libllbuild/BuildSystem-C-API.cpp:475]: Agree with @owenv, despite the verbosity, forced casts based on assumptions about the underlying integer value can be dangerous.

File: products/libllbuild/include/llbuild/buildsystem.h [hunk 4]
```diff
@@ -203,6 +203,24 @@ typedef enum LLBUILD_ENUM_ATTRIBUTES {
     llb_quality_of_service_unspecified = 4
 } llb_quality_of_service_t LLBUILD_SWIFT_NAME(QualityOfService);
 
+/// Rule run reasons.
+typedef enum LLBUILD_ENUM_ATTRIBUTES {
+    /// The rule has never been built.
+    llb_rule_run_reason_never_built LLBUILD_SWIFT_NAME(neverBuilt) = 0,
```
Comment (jakepetroules) [inline products/libllbuild/include/llbuild/buildsystem.h:209]: LLBUILD_SWIFT_NAME 👍

**Review by jakepetroules — APPROVED** [review]

**Comment by jakepetroules** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
It turns out tests sometimes use a BuildSystemDelegate which is not a BuildSystemFrontendDelegate, so the `static_cast<BuildSystemFrontendDelegate*>(&getBuildSystem().getDelegate())` was unsafe. I've added `determinedRuleNeedsToRun` as an abstract method on BuildSystemDelegate, which lets me drop the unsound cast and generally seems reasonable.

**Comment by owenv** [comment]
@swift-ci please smoke test and merge

**Comment by owenv** [comment]
@swift-ci please smoke test

**Pull request merged**